File: Src/Mercury.Messaging.Test/Agents/Mock/Paddle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Mercury.Messaging.Core;
using Mercury.Messaging.Runtime;

namespace Mercury.Messaging.Test.Agents.Mock
{
    public class Paddle : Agent
    {
        public Paddle(AgentPort port)
        {
            port.Receive<Ping>((msg) => 
                {
                    if (this.Count < 3)
                    {
                        IncrementCount();
                        port.Environment.RoutingEngine.Send<Pong>(msg.Instance, new Pong(port.Id));
                    }
                    else
                    {
                        port.Environment.RoutingEngine.Send<PongFinal>(msg.Instance, new PongFinal());
                        port.Environment.Kill(port.Id);
                    }
                });

            port.Receive<Pong>((msg) => 
                {
                    port.Environment.RoutingEngine.Send<Ping>(msg.Instance, new Ping(port.Id));
                });

            port.Receive<PongFinal>((msg) => 
                {
                    port.Environment.Kill(port.Id);
                });
        }

        private int _count;
        public int Count 
        { 
            get { return this._count; } 
        }

        protected void IncrementCount()
        {
            System.Threading.Interlocked.Add(ref this._count, 1);
        }
    }
}

File: Src/Samples/Samples.AgentPingPong/Agents/Paddle.vb
Imports Mercury.Messaging.Core
Imports Mercury.Messaging.Runtime
Imports Mercury.Messaging.Messages

''' <summary>
''' Represents a paddle used return a ping or pong.
''' </summary>
''' <remarks></remarks>
Public Class Paddle
    Implements Agent

    Public Sub New(ByVal port As AgentPort)
        Dim environment As RuntimeEnvironment = port.Environment

        port.Receive(Of Ping)(
            Sub(msg)
                Dim sender As LocalRef = environment.GetRef(msg.SenderId)
                If (msg.Counter = 0) Then
                    sender.Send(Of [Stop])(New [Stop]())
                    Console.Write(String.Format("Shutdown:{0}", port.Id))
                    port.Shutdown(Nothing)
                Else
                    Dim counter As Integer = msg.Counter - 1
                    Console.WriteLine(String.Format("{0} received Ping {1}. Sent Pong.", port.Id, msg.Counter))
                    sender.Send(Of Pong)(New Pong(port.Id, counter))
                End If
            End Sub)

        port.Receive(Of Pong)(
            Sub(msg)
                Dim sender As LocalRef = environment.GetRef(msg.SenderId)
                'Console.WriteLine(String.Format("{0} received Pong {1}.", port.Id, msg.Counter))
                Console.WriteLine(String.Format("Received Pong. Sent Ping.", port.Id))
                sender.Send(Of Ping)(New Ping(port.Id, msg.Counter))
            End Sub)
    End Sub
End Class